src/nvim/os/fsnotification.c

```diff
@@ -0,0 +1,248 @@
+{
+  assert(watchers_list != NULL);
+
+  kliter_t(Watcher)* it = kl_begin(watchers_list);
+
+  for (size_t i = 0; i < watchers_list->size; i++) {
+    if (kl_val(it).buffer == buffer) {
+      if (w != NULL)
+        *w = kl_val(it);
+      return true;
+    }
+
+    it = kl_next(it);
+  }
+
+  return false;
+}
+
+static bool remove_watcher(buf_T* buffer, Watcher* w)
+{
+  assert(watchers_list != NULL);
+
+  kliter_t(Watcher)* it = NULL;
+  if (find_node_before(watchers_list, buffer, &it)) {
+    if (it == NULL) {  // the element is the head, use shift
+      kl_shift(Watcher, watchers_list, w);
+    } else {
+      kl_iter_remove_next(Watcher, watchers_list, it, w);
+    }
+
+    return true;
+  }
+
+  return false;
+}
+
+void fsnotification_init(void) {
+  watchers_list = kl_init(Watcher);
+  event_lookup = kh_init(EventTable);
+}
+
+// called on application exiting
+void fsnotification_teardown(void) {
+  Watcher w;
+
+  while (kl_shift(Watcher, watchers_list, &w) == 0) {
+    free_watcher(&w);
+  }
+
+  kl_destroy(Watcher, watchers_list);
+  kh_destroy(EventTable, event_lookup);
+}
+
+bool fsnotification_add_buffer(buf_T *buffer) {
+  // We check first if the buffer is already register
+  // In that case, we just enable its watcher
+  {
+    Watcher w;
+    if (find_watcher(buffer, &w)) {
+     return fsnotification_enable_watcher(&w, true);
+    }
+  }
+
+  uv_fs_event_t *event_handle = try_malloc(sizeof(uv_fs_event_t));
+
+  if (event_handle == NULL) {  // malloc failed
+    return false;
+  }
+
+  if (uv_fs_event_init(uv_default_loop(), event_handle) != 0) {
+    return false;
+  }
+
+  Watcher w = {
+    .buffer = buffer,
+    .handle = event_handle,
+  };
+
+  assert(watchers_list != NULL);
+  Watcher* new_element = kl_pushp(Watcher, watchers_list);
+  *new_element = w;
+
+  return fsnotification_enable_watcher(new_element, true);
+}
+
+static bool fsnotification_enable_watcher(Watcher* watcher, bool state)
+{
+  if (uv_is_active((uv_handle_t*) watcher->handle) != state) {
+    int res;
+    if (state) {
+      res = uv_fs_event_start(
+            watcher->handle,
+            fs_event_callback,
+            (const char*)watcher->buffer->b_ffname,
+            UV_FS_EVENT_RECURSIVE);
+
+    } else {
+      res = uv_fs_event_stop(watcher->handle);
+    }
+
+    return res == 0;  // success
+  }
+
+  // success if the final state is the desired one
+  return true;
+}
+
+void fsnotification_delete_buffer(buf_T* buffer) {
+  Watcher w;
+  if (remove_watcher(buffer, &w)) {
+    free_watcher(&w);
+  }
+
+  khiter_t value = kh_get(EventTable, event_lookup, (char*)buffer->b_ffname);
+  if (value != kh_end(event_lookup))
+  {
+    kh_del(EventTable, event_lookup, value);
+  }
+}
+
+bool fsnotification_enable_watcher_on(buf_T* buffer, bool state) {
+  Watcher w;
+  if (find_watcher(buffer, &w)) {
+    return fsnotification_enable_watcher(&w, state);
+  }
+
+  return false;
+}
+
+static void free_watcher(Watcher* watcher) {
```

Comment:
`{` should be on its own line. Also missing documentation (I do not see documentation near any function) and function attributes (in this case FUNC_ATTR_NONNULL_ALL).